diff --git a/cjs/src/index.js b/cjs/src/index.js
--- a/cjs/src/index.js
+++ b/cjs/src/index.js
@@ -241,12 +241,18 @@
     let savepoints = 0
       , connection
       , prepare = null
+      , closedError = null
 
     try {
       await sql.unsafe('begin ' + options.replace(/[^a-z ]/ig, ''), [], { onexecute }).execute()
       return await Promise.race([
         scope(connection, fn),
-        new Promise((_, reject) => connection.onclose = reject)
+        new Promise((_, reject) => connection.onclose = error => {
+          closedError = error
+          while (queries.length)
+            queries.shift().reject(error)
+          reject(error)
+        })
       ])
     } catch (error) {
       throw error
@@ -304,6 +310,9 @@
       }
 
       function handler(q) {
+        if (closedError)
+          return q.reject(closedError)
+
         q.catch(e => uncaughtError || (uncaughtError = e))
         c.queue === full
           ? queries.push(q)
diff --git a/src/index.js b/src/index.js
--- a/src/index.js
+++ b/src/index.js
@@ -238,12 +238,18 @@
     let savepoints = 0
       , connection
       , prepare = null
+      , closedError = null
 
     try {
       await sql.unsafe('begin ' + options.replace(/[^a-z ]/ig, ''), [], { onexecute }).execute()
       return await Promise.race([
         scope(connection, fn),
-        new Promise((_, reject) => connection.onclose = reject)
+        new Promise((_, reject) => connection.onclose = error => {
+          closedError = error
+          while (queries.length)
+            queries.shift().reject(error)
+          reject(error)
+        })
       ])
     } catch (error) {
       throw error
@@ -301,6 +307,9 @@
       }
 
       function handler(q) {
+        if (closedError)
+          return q.reject(closedError)
+
         q.catch(e => uncaughtError || (uncaughtError = e))
         c.queue === full
           ? queries.push(q)
